### PR TITLE
Use WKDataTask for SystemPreview downloads

### PR DESCRIPTION
**Pull request by grorg**
#### d55e3cae6649895b6d2aeb7235ba879bdb403868
<pre>
Use WKDataTask for SystemPreview downloads
<a href="https://bugs.webkit.org/show_bug.cgi?id=255954">https://bugs.webkit.org/show_bug.cgi?id=255954</a>
rdar://108524222

Reviewed by Tim Horton.

Migrate away from LegacyDownloadClient and instead handle System Previews
directly. This will allow us to remove the horrible code in the download
client, as well eventually allow 3rd-party engines to use ARQL.

The fix is to intercept the HTMLAnchorElement click for a system
preview, and send that directly to the SystemPreviewController in
the UI Process. That object then creates a WKDataTask for the download
and presents ARQL as normal.

A followup patch will remove the old code.

* Source/WebCore/html/HTMLAnchorElement.cpp:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/page/Page.cpp:
* Source/WebCore/page/Page.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
* Source/WebKit/UIProcess/Cocoa/SystemPreviewControllerCocoa.mm:
* Source/WebKit/UIProcess/SystemPreviewController.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SystemPreview.mm:
(-[TestSystemPreviewUIDelegate _presentingViewControllerForWebView:]):
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnitBox.usdz: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/system-preview-trigger.html:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/system-preview.html: Added.

Canonical link: <a href="https://commits.webkit.org/263393@main">https://commits.webkit.org/263393@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/65b5f699d96e45b9b21b3361fce847106c3f8401

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4481 "2 style errors") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/4597 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4742 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5964 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/4660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/4469 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/4729 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/4555 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/4892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/4546 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/4729 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/4015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5968 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/23/builds/4729 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/4000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/8560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/4729 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/4070 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/5603 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/4470 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/4555 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/3998 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/3999 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/8045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/513 "Failed to push commit to Webkit repository") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/4359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | |

### CODE REVIEW

File: Source/WebKit/UIProcess/Cocoa/SystemPreviewControllerCocoa.mm [hunk 1]
```diff
@@ -253,8 +257,216 @@ - (UIImage *)previewController:(QLPreviewController *)controller transitionImage
...
+- (void)dataTask:(_WKDataTask *)dataTask didReceiveData:(NSData *)data
+{
+    ASSERT(_data);
+    if (_data.get().length + data.length > SystemPreviewMaximumSize) {
+        [dataTask cancel];
+        _previewController->loadFailed();
+        return;
+    }
+
+    [_data.get() appendData:data];
```
Comment (hortont424) [inline Source/WebKit/UIProcess/Cocoa/SystemPreviewControllerCocoa.mm]: no .get()

File: Source/WebKit/UIProcess/Cocoa/SystemPreviewControllerCocoa.mm [hunk 2]
```diff
@@ -59,6 +61,8 @@ + (void)launchPreviewApplicationWithURLs:(NSArray *)urls completion:(void (^)(NS
 
 static NSString * const _WKARQLWebsiteURLParameterKey = @"ARQLWebsiteURLParameterKey";
 
+const int64_t SystemPreviewMaximumSize = 40 * 1024 * 1024;
```
Comment (hortont424) [inline Source/WebKit/UIProcess/Cocoa/SystemPreviewControllerCocoa.mm]: This seems awfully tiny

File: Source/WebKit/UIProcess/Cocoa/SystemPreviewControllerCocoa.mm [hunk 3]
```diff
@@ -253,8 +257,216 @@ - (UIImage *)previewController:(QLPreviewController *)controller transitionImage
...
+    [self completeLoad];
+}
+
+- (void)completeLoad
+{
+    FileSystem::PlatformFileHandle fileHandle;
+    auto filePath = FileSystem::openTemporaryFile("SystemPreview"_s, fileHandle, ".usdz"_s);
+    ASSERT(FileSystem::isHandleValid(fileHandle));
+
+    size_t byteCount = FileSystem::writeToFile(fileHandle, _data.get().bytes, _data.get().length);
```
Comment (hortont424) [inline Source/WebKit/UIProcess/Cocoa/SystemPreviewControllerCocoa.mm]: I would [_data bytes] (etc.), but I know some people like this instead?

**Review by hortont424 — APPROVED** [review]

File: Source/WebKit/UIProcess/Cocoa/SystemPreviewControllerCocoa.mm (hunk 2 above)
Comment (grorg) [reply Source/WebKit/UIProcess/Cocoa/SystemPreviewControllerCocoa.mm]: I will remove the limit.

File: Source/WebKit/UIProcess/Cocoa/SystemPreviewControllerCocoa.mm [hunk 4]
```diff
@@ -253,8 +257,216 @@ - (UIImage *)previewController:(QLPreviewController *)controller transitionImage
 
 @end
 
+@interface _WKSystemPreviewDataTaskDelegate : NSObject <_WKDataTaskDelegate> {
+    WebKit::SystemPreviewController* _previewController;
+    long long _expectedContentLength;
+    long long _receivedLength;
```
Comment (pxlcoder) [inline Source/WebKit/UIProcess/Cocoa/SystemPreviewControllerCocoa.mm]: I don't see this variable used anywhere?

**Pull request merged**